include/podio/detail/OrderKey.h

```diff
@@ -0,0 +1,34 @@
+#ifndef PODIO_DETAIL_ORDERKEY_H
+#define PODIO_DETAIL_ORDERKEY_H
+#include <cstdint>
+
+namespace podio::detail {
+// Internal class allowing datatype objects to be placed in data structures like maps and sets by providing a way to
+// compare them. The comparison is based on a numeric value derived from the address of their internal data objects.
+//
+// This class is intended to be used as the return value of internal `podio::detail::getOrderKey` free functions. These
+// functions are friends of each datatype, allowing them to access the internal data objects and define less-than
+// comparison operators for both datatypes and interface types.
+//
+// The friend free function design is used in order to reduce the coupling between interfaces and datatypes. Interfaces
+// do not need to be friends of datatypes to define the less-than comparison operator, which allows using datatypes from
+// different datamodels in an interface type.
+class OrderKey {
+public:
+  OrderKey(void* orderKey) : m_orderKey(reinterpret_cast<value_type>(orderKey)) {
+  }
+  friend bool operator<(const OrderKey& lhs, const OrderKey& rhs) {
+    return lhs.m_orderKey < rhs.m_orderKey;
+  }
+
+private:
+#ifdef UINTPTR_MAX
+  using value_type = uintptr_t;
+#else
+  using value_type = uintmax_t;
+#endif
+  value_type m_orderKey;
```

Comment:
```suggestion
  void* m_orderKey;
```

Just to have a papertrail of the [EDM4hep meeting today](https://indico.cern.ch/event/1484070/)

Reply:
Thanks, I'm having second thoughts on this now

It's super pedantic, but the standard guarantees implementation defined strict-ordering of all the pointers for comparisons done with `std::less` - comparing two pointers to unrelated objects with plain `<` yields 'unspecified' result. Taking look at the `std::less` for pointers from libstdc++ it does cast the pointers to internal numeric type then compare them with `<`.

So I think it's marginally better to either keep the casting or go with `less` rather than do plain comparison for pointers. On the other hand so far we had plain comparisons and haven't seen any problems yet

Reply:
What are "unrelated objects" in this case? One way to exclude undefined behavior would be to mark the `operator<` as `constexpr`. In that case the compiler should refuse to compile it, I think. (Also we do have builds with *UBSan* and either that isn't able to flag this, or it's fine?)

If `std::less` solves all of this (with our without `void*`), I would just go for that here.

Reply:
If I understood [cppreference](https://en.cppreference.com/w/cpp/language/operator_comparison) ('Built-in pointer relational comparison' section) correctly, unspecified occurs unless the pointers are from the same array, or a pointer is from an array and another is past the array, or they are pointers to members of the same object

Yeah, let's do `std::less` and don't dwell on this